src/violation.rs

```diff
@@ -1,15 +1,98 @@
-// TODO: Implement Violations for constraints
-#[derive(Debug, Clone)]
-pub struct Violations {
-    violations: Vec<Violation>,
+/// Represents Violations found during validation of values based on schema types
+pub trait Violations {
+    /// adds a violation into the collection of [Violation]s
+    fn add_violation(&mut self, violation: ViolationImpl);
+
+    /// Returns `true` if no violations were found, otherwise `false`
+    fn is_valid(&self) -> bool;
+
+    /// Returns the [Violation]s that were found during validation
+    fn violations(&self) -> &[ViolationImpl];
+
+    /// Creates a checkpoint to track the number of violations from this point onwards
+    fn checkpoint(&self) -> Checkpoint;
+}
+
+// Represents collection of [Violation]s
+pub struct ViolationsImpl {
+    violations: Vec<ViolationImpl>,
+}
+
+impl ViolationsImpl {
+    pub fn new() -> Self {
+        Self { violations: vec![] }
+    }
 }
 
-impl Violations {
-    fn add_violation(violation: Violation) {
-        todo!()
+impl Violations for ViolationsImpl {
+    fn add_violation(&mut self, violation: ViolationImpl) {
+        self.violations.push(violation);
     }
+
+    fn is_valid(&self) -> bool {
+        self.violations.is_empty()
+    }
+
+    fn violations(&self) -> &[ViolationImpl] {
+        &self.violations
+    }
+
+    fn checkpoint(&self) -> Checkpoint {
```

Comment:
Why not making this `fn len(&self) -> usize` for simplicity?

EDIT: I see below that this is a placeholder for more complicated logic.